batchflow/opensets/small_imagenet.py

```diff
@@ -0,0 +1,138 @@
+""" Contains Imagenette and Imagewoof datasets """
+
+import os
+from os.path import dirname, basename
+import tempfile
+import logging
+import urllib.request
+import tarfile
+from io import BytesIO
+
+import PIL
+import tqdm
+import numpy as np
+from sklearn.preprocessing import LabelEncoder
+
+from . import ImagesOpenset
+from .. import DatasetIndex
+
+
+logger = logging.getLogger('SmallImagenet')
+
+
+class SmallImagenet(ImagesOpenset):
+    """ Base class for ImagenetLike datasets
+
+    Notes
+    -----
+    - Datasets contain both grayscale and colored images.
+      Argument `drop_grayscale` controls whether grayscale images should be dropped.
+
+    """
+    SOURCE_URL = None
+    num_classes = 10
+
+    def __init__(self, *args, drop_grayscale=True, bar=False, preloaded=None, train_test=True, **kwargs):
+        self.bar = tqdm.tqdm(total=2) if  bar else None
+        self.drop_grayscale = drop_grayscale
+        super().__init__(*args, preloaded=preloaded, train_test=train_test, **kwargs)
+        if self.bar:
+            self.bar.close()
+
+    def download(self, path=None):
+        """ Load data from website and extract it into numpy arrays """
+
+        def _image_class(filepath):
+            """ Image's class is determined by the parent folder of the image """
+            return basename(dirname(filepath))
+
+        def _is_train(filepath):
+            """ Whether image belongs to train or val parts can be determined by
+            the level 2 parent folder of the image
+            """
+            return basename(dirname(dirname(filepath))) == 'train'
+
+        def _extract(archive_file, member):
+            data = archive_file.extractfile(member).read()
+            return PIL.Image.open(BytesIO(data))
+
+        def _gather_extracted(archive, files):
+            images = np.array([_extract(archive, file) for file in files], dtype=object)
+            labels = np.array([_image_class(file.name) for file in files])
+            labels_encoded = LabelEncoder().fit_transform(labels)
+
+            if self.drop_grayscale:
+                rgb_mask = [i for i in range(len(images)) if images[i].mode == 'RGB']
+                return images[rgb_mask], labels_encoded[rgb_mask]
```

Comment:
This requires a full copy of data which takes a lot of time and resources.
Another problem is that when they fix grayscale images this becomes an awful resource waste with no reason whatsoever.